applications/tari_dan_wallet_daemon/src/webrtc.rs

```diff
@@ -0,0 +1,147 @@
+
+use std::sync::{Arc, Mutex};
+
+use anyhow::Result;
+use axum_jrpc::{JsonRpcAnswer, JsonRpcResponse};
+use reqwest::header::CONTENT_TYPE;
+use serde::Deserialize;
+use tari_shutdown::ShutdownSignal;
+use webrtc::{
+    api::APIBuilder,
+    data_channel::{data_channel_message::DataChannelMessage, RTCDataChannel},
+    ice_transport::{
+        ice_candidate::{RTCIceCandidate, RTCIceCandidateInit},
+        ice_server::RTCIceServer,
+    },
+    peer_connection::{configuration::RTCConfiguration, sdp::session_description::RTCSessionDescription},
+};
+
+#[derive(Deserialize, Debug)]
+struct Offer {
+    offer: String,
+}
+
+pub async fn handle_data(msg: String) -> Result<String> {
+    println!("webrtc received {msg}");
+    let (method, params) = match msg.split_once(' ') {
+        Some(a) => a,
+        None => (msg.as_str(), ""),
+    };
+    let url = "http://127.0.0.1:9000";
+    let client = reqwest::Client::new();
+    let body = format!(
+        "{{\"method\":\"{}\", \"jsonrpc\":\"2.0\", \"id\": 1, \"params\":[{}]}}",
+        method, params
+    );
+    println!("Request {body}");
+    let resp = client
+        .post(url)
+        .header(CONTENT_TYPE, "application/json")
+        .body(body)
+        .send()
+        .await?
+        .json::<JsonRpcResponse>()
+        .await?;
+    match resp.result {
+        JsonRpcAnswer::Result(result) => Ok(result.to_string()),
+        JsonRpcAnswer::Error(error) => Ok(error.to_string()),
+    }
+}
+
+fn get_rtc_configuration() -> RTCConfiguration {
+    RTCConfiguration {
+        ice_servers: vec![RTCIceServer {
+            urls: vec!["stun:stun.l.google.com:19302".to_owned()],
+            ..Default::default()
+        }],
+        ..Default::default()
+    }
+}
+
+pub async fn webrtc_start_session(session: String, shutdown_signal: ShutdownSignal) -> Result<()> {
+    let api = APIBuilder::new().build();
+
+    let pc = api.new_peer_connection(get_rtc_configuration()).await?;
+    pc.on_data_channel(Box::new(move |d: Arc<RTCDataChannel>| {
+        Box::pin(async move {
+            let d_on_message = d.clone();
+            d.on_message(Box::new(move |msg: DataChannelMessage| {
+                let d_on_message = d_on_message.clone();
+                Box::pin(async move {
+                    let msg_str = String::from_utf8(msg.data.to_vec()).unwrap();
+                    d_on_message
+                        .send_text(handle_data(msg_str).await.unwrap())
+                        .await
+                        .unwrap();
+                })
+            }))
+        })
+    }));
+
+    let session_clone = session.clone();
+
+    let ice_cnt = Arc::new(Mutex::new(0));
+
+    pc.on_ice_candidate(Box::new(move |d: Option<RTCIceCandidate>| {
+        let url;
+        {
+            let mut id = ice_cnt.lock().unwrap();
+            url = format!("https://webrtc-test-d081d-default-rtdb.europe-west1.firebasedatabase.app/{session_clone}/answer_ice/{}.json",*id);
```

Comment:
should this be hardcoded ?

Reply:
Currently yes. We have to come up with our signaling server. This is not a signaling server, it's just a real database, so the interface will be different. So the code will change as well, not just the address.